### PR TITLE
[BEAM-1297] Update maven plugin versions and fix some maven related issues

### PR DESCRIPTION
**Pull request by iemejia**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by iemejia** [comment]
R: @jbonofre

**Comment by iemejia** [comment]
As mentioned in the JIRA, I was trying to make some built speed ups and I found some issues + upgrade the versions of the maven plugins. This is from the initial report via:
```
mvn versions:display-plugin-updates
```

**Comment by iemejia** [comment]
Ok I found a last minute mistake, I just rebased the fix, It should be ok now.

**Comment by iemejia** [comment]
A rebase after the beam version upgrade to 0.6.0.

File: .jenkins/common_job_properties.groovy [hunk 1]
```diff
@@ -138,8 +138,11 @@ class common_job_properties {
   // Sets common config for Maven jobs.
   static def setMavenConfig(def context) {
     context.mavenInstallation('Maven 3.3.3')
+    context.mavenOpts('-client')
     context.mavenOpts('-Dorg.slf4j.simpleLogger.showDateTime=true')
     context.mavenOpts('-Dorg.slf4j.simpleLogger.dateTimeFormat=yyyy-MM-dd\\\'T\\\'HH:mm:ss.SSS')
+    context.mavenOpts('-XX:+TieredCompilation')
+    context.mavenOpts('-XX:TieredStopAtLevel=1')
     context.rootPOM('pom.xml')
```
Comment (jbonofre) [inline .jenkins/common_job_properties.groovy]: What's the purpose of these changes ?

File: .travis.yml [hunk 2]
```diff
@@ -49,7 +49,7 @@ matrix:
       env: MAVEN_OVERRIDE="-Peclipse-jdt -DskipTests $MAVEN_OVERRIDE $MAVEN_CONTAINER_OVERRIDE" CUSTOM_JDK="oraclejdk8"
 
 before_install:
-  - echo 'MAVEN_OPTS="$MAVEN_OPTS -Xmx1024m -XX:MaxPermSize=512m -XX:+BytecodeVerificationLocal"' >> ~/.mavenrc
+  - echo 'MAVEN_OPTS="$MAVEN_OPTS -client -Xmx1024m -XX:MaxPermSize=512m -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -XX:+BytecodeVerificationLocal"' >> ~/.mavenrc
   - echo $'MAVEN_OPTS="$MAVEN_OPTS -Dorg.slf4j.simpleLogger.showDateTime=true -Dorg.slf4j.simpleLogger.dateTimeFormat=yyyy-MM-dd\'T\'HH:mm:ss.SSS"' >> ~/.mavenrc
```
Comment (jbonofre) [inline .travis.yml]: Same question here.

File: examples/java/pom.xml [hunk 3]
```diff
@@ -327,24 +327,21 @@
...
+      <!-- BEAM-933 -->
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>findbugs-maven-plugin</artifactId>
+        <configuration>
+          <skip>true</skip>
+        </configuration>
+      </plugin>
+
       <plugin>
```
Comment (jbonofre) [inline examples/java/pom.xml]: Superfluous change.

File: examples/java8/pom.xml [hunk 4]
```diff
@@ -142,6 +129,15 @@
...
+      <!-- BEAM-934 -->
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>findbugs-maven-plugin</artifactId>
+        <configuration>
+          <skip>true</skip>
+        </configuration>
+      </plugin>
+
       <plugin>
```
Comment (jbonofre) [inline examples/java8/pom.xml]: Same here.

File: pom.xml [hunk 5]
```diff
@@ -1257,7 +1258,7 @@
       <plugin>
         <groupId>org.codehaus.mojo</groupId>
         <artifactId>versions-maven-plugin</artifactId>
-        <version>2.1</version>
+        <version>2.3</version>
         <reportSets>
```
Comment (jbonofre) [inline pom.xml]: Just curious: did you check the resulting artifacts ? It looks good to me anyway.

File: runners/spark/pom.xml [hunk 6]
```diff
@@ -409,7 +322,84 @@
...
+                  <pattern>com.google.thirdparty</pattern>
+                  <shadedPattern>org.apache.beam.spark.relocated.com.google.thirdparty</shadedPattern>
+                </relocation>
+              </relocations>
+              <transformers>
+                <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer" />
+              </transformers>
+            </configuration>
+          </execution>
+        </executions>
```
Comment (jbonofre) [inline runners/spark/pom.xml]: Again, it creates "noise" during the review. You just change the location of this section right ?

File: .jenkins/common_job_properties.groovy (hunk 1 above)
Comment (iemejia) [inline .jenkins/common_job_properties.groovy]: By default the JVM on linux runs in server mode and optimizes the JIT for long running processes (it records more statistics about the execution to profile things afterwards). Maven processes are not long running processes so this optimization costs some extra time when you start the JVM.
The idea comes from point 5 here, and if you test it you can see the difference.
https://zeroturnaround.com/rebellabs/your-maven-build-is-slow-speed-it-up/

Note however that the produced code is not different because this changes the options of the JVM for the maven execution not for the java compiler (I compared the generated classes with and without this flags to see if there were differences).

File: examples/java/pom.xml (hunk 3 above)
Comment (iemejia) [inline examples/java/pom.xml]: My rational is that pluginManagement is intended to configure project builds that inherit from this one (it creates a context) but in the files where I changed this, those are 'leaf' poms not parents so this is not needed. I agree that this can be a subtle difference but my motivation was also to get rid of the warnings that this produced on unsetted version for 'fake parents'.

File: pom.xml (hunk 5 above)
Comment (iemejia) [inline pom.xml]: Yes, for execution, however I did not do the equivalent of a full apache release because I don't know how to test this, but the jars build and the tests execute correctly.

File: runners/spark/pom.xml (hunk 6 above)
Comment (iemejia) [inline runners/spark/pom.xml]: Actually all the relocations are isolated in one commit but this is hard to view in the default github interface, if you consider that this is not ok, I can get rid of this commit, and don't touch the pluginManagement, but I think the leaf argument (see above) makes sense.

**Comment by iemejia** [comment]
Just answered comments, please tell me if you want any change and I will do it and rebase the thing.

**Comment by iemejia** [comment]
Oh and thanks for the review!

**Comment by iemejia** [comment]
@dhalperi I am looking for someone to take the review/merge of this one since JB is on holidays, can you (or someone else) take a look Thx.

File: examples/java/pom.xml (hunk 3 above)
Comment (dhalperi) [inline examples/java/pom.xml]: The current pattern, I think, was designed for `pom.xml` being as easy to skim as possible. The files are so long... We currently have (or aim for) a pattern that

1. all configuration appears in `pluginManagement`,
1. you look at the `plugins` section just to see which plugins are uniquely/newly active in this module and its children.

What do you think?

Separately, can you comment on the warnings?

File: pom.xml [hunk 7]
```diff
@@ -979,7 +979,7 @@
 
         <plugin>
           <artifactId>maven-compiler-plugin</artifactId>
-          <version>3.5.1</version>
```
Comment (dhalperi) [inline pom.xml]: no strong objection, but any specific need for these changes?

File: pom.xml [hunk 8]
```diff
@@ -1298,7 +1299,7 @@
         <plugin>
           <groupId>org.codehaus.mojo</groupId>
           <artifactId>findbugs-maven-plugin</artifactId>
-          <version>${findbugs.version}</version>
```
Comment (dhalperi) [inline pom.xml]: revert? or remove `findbugs.version` property if this is the only place it's used?

File: .jenkins/common_job_properties.groovy [hunk 9]
```diff
@@ -162,8 +162,11 @@ class common_job_properties {
   // Sets common config for Maven jobs.
```
Comment (dhalperi) [inline .jenkins/common_job_properties.groovy]: if you don't mind, I'd pull this into its own PR. The rest are almost no-ops, this one seems like a behavior change that might need more thought/review.

File: examples/java/pom.xml (hunk 3 above)
Comment (iemejia) [inline examples/java/pom.xml]: Ok, I didn't get the logic for this artificial pluginManagement sections on the leaves, I think it makes sense from a logic separation point of view, however the documentation does not say that the use is intended to configure plugins at the same hierarchy level.
"Plugin Management contains plugin elements in much the same way, except that rather than configuring plugin information for this particular project build, it is intended to configure project builds that inherit from this one. However, this only configures plugins that are actually referenced within the plugins element in the children."

Comment (iemejia) [inline examples/java/pom.xml]: I was looking at the warning and it is not critical, it says that there is not a defined plugin version for the plugins that are not inherited (the case of all the uses in this commit).
If you think it is better I will remove that commit, to reduce the possible impact of the changes and keep the current structure.

File: pom.xml (hunk 7 above)
Comment (iemejia) [inline pom.xml]: No, special need, the goal is to keep up to date the maven plugins, in particular those who can have performance improvements for the build.

File: pom.xml [hunk 10]
```diff
@@ -1430,4 +1430,8 @@
       </plugin>
     </plugins>
   </reporting>
+
+  <prerequisites>
+    <maven>3.2</maven>
+  </prerequisites>
```
Comment (iemejia) [inline pom.xml]: No, this one was a huge surprise for me, the maven plugin resolution mechanism is quite curious, they assume by default that users have the lowest version possible of maven and resolve the plugin versions from it, in a really conservative manner, it looks for the lowest version of the plugin for those that are not explicit. This can seem redundant but actually it forces the plugin resolution to consider that the user has in the worst case the version defined here.

File: .jenkins/common_job_properties.groovy (hunk 9 above)
Comment (iemejia) [inline .jenkins/common_job_properties.groovy]: I agree with you, probably the wisest is to do this changes incrementally, so I will rebase this, so we can merge first the version changes, and then the other ones if you agree.

**Comment by iemejia** [comment]
Ok @dhalperi I rebased as I proposed, tell me if it is ok now. I will create a new PR for the OPs changes. Notice that the missing changes are the one who bring the performance improvements.

File: pom.xml (hunk 10 above)
Comment (dhalperi) [inline pom.xml]: Okay. Would you mind adding a comment in both places to change the other if making changes? :)

Comment (dhalperi) [inline pom.xml]: Made this fixup in the merge. Thanks Ismaël!

**Comment by iemejia** [comment]
I will submit the rest of the PR next week so we can have some time to fix any possible issue with the plugins versions. Thanks for the review @jbonofre and @dhalperi.
